### PR TITLE
 jwt_authn filter config with ECDS

### PR DESCRIPTION
**Pull request by aryan16**
**Please provide a description of this PR:**

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [x] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/extension_config_builder.go [hunk 1]
```diff
@@ -33,3 +34,17 @@ func (configgen *ConfigGeneratorImpl) BuildExtensionConfiguration(
 	extensions = append(extensions, extension.InsertedExtensionConfigurations(wasmPlugins, extensionConfigNames, pullSecrets)...)
 	return extensions
 }
+
+func (configgen *ConfigGeneratorImpl) BuildExtensionConfigurationForJwks(proxy *model.Proxy,
+	req *model.PushRequest, env *model.Environment) *core.TypedExtensionConfig {
+	req.Push.InitContext(env, req.Push, req)
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/extension_config_builder.go:40]: this shouldn't be needed, it should be called higher up

File: pilot/pkg/xds/ecds.go [hunk 2]
```diff
@@ -64,7 +66,14 @@ func (e *EcdsGenerator) Generate(proxy *model.Proxy, w *model.WatchedResource, r
...
+	jwtEcds := e.Server.ConfigGenerator.BuildExtensionConfigurationForJwks(proxy, req, e.Server.Env)
+	if jwtEcds != nil {
+		resources = append(resources, &discovery.Resource{
+			Name:     jwtEcds.Name,
+			Resource: protoconv.MessageToAny(jwtEcds),
+		})
+	}
 	secretResources := referencedSecrets(proxy, req.Push, w.ResourceNames)
 	// Check if the secret updates is relevant to Wasm image pull. If not relevant, skip pushing ECDS.
 	if !model.ConfigsHaveKind(req.ConfigsUpdated, kind.WasmPlugin) && !model.ConfigsHaveKind(req.ConfigsUpdated, kind.EnvoyFilter) &&
```
Comment (howardjohn) [inline pilot/pkg/xds/ecds.go:79]: fyi: this part will need to be updated

**Comment by kyessenov** [comment]
Cool. Want to make sure one aspect is done right: ECDS reload does not terminate existing long streams. We generally relied on listener reload before ECDS. It can be done in Envoy, but we need to get the right requirements.

@ingwonsong Actually, same problem exists with authn Wasm filters?

**Comment by ingwonsong** [comment]
> Cool. Want to make sure one aspect is done right: ECDS reload does not terminate existing long streams. We generally relied on listener reload before ECDS. It can be done in Envoy, but we need to get the right requirements.
> 
> @ingwonsong Actually, same problem exists with authn Wasm filters?

I am not sure that I correctly understood your question, but Wasm issue was related that Envoy may not create ECDS request if the corresponding ECDS is already delivered before.

**Comment by kyessenov** [comment]
@ingwonsong Let me elaborate. The issue is that Wasm API does not retroactively apply to on-going requests or connections. When it comes to authn/authz, there is an expectation of certain deadline when policies are enforced, and this is a violation.

**Comment by ingwonsong** [comment]
ah.. yes. it is. we may need to introduce TTL for ECDS resources.

**Comment by kyessenov** [comment]
@ingwonsong I don't think TTL helps here. The issue is that we need to reapply new config to old connections. Because connections are not drained when ECDS updates, the old connection continues using old ECDS forever.

The current mitigation is to update the entire LDS which is expensive.

**Comment by ingwonsong** [comment]
Ah. I got your point. Thanks for the detailed explanation.

**Pull request closed without merging**